domain/application/service/charm.go

```diff
@@ -609,6 +616,16 @@ func (s *Service) GetCharmDownloadInfo(ctx context.Context, id corecharm.ID) (*c
 	return s.st.GetCharmDownloadInfo(ctx, id)
 }
 
+// GetAvailableCharmArchiveSHA256 returns the SHA256 hash of the charm archive
+// for the given charm id. If the charm is not available,
+// [applicationerrors.CharmNotResolved] is returned.
+func (s *Service) GetAvailableCharmArchiveSHA256(ctx context.Context, id corecharm.ID) (string, error) {
+	if err := id.Validate(); err != nil {
+		return "", fmt.Errorf("charm id: %w", err)
+	}
+	return s.st.GetAvailableCharmArchiveSHA256(ctx, id)
+}
```

Comment:
I think @nvinuesa is proposing something very similar in one of his patches

Reply:
Indeed. What Jack's referring to is being done in https://github.com/juju/juju/pull/18504/files#diff-f634e196c8c372b1bc36ad4d6db384d26713e39a92b90c981816480a2d1e91d6R494-R510.
I've done it a bit different, by reusing the `GetCharmArchiveMetadata` method from State which already returns the hash. I think you should too, instead of creating a new state method just for the hash:
```
	_, hash, err := s.st.GetCharmArchiveMetadata(ctx, id)
	if err != nil {
		return "", internalerrors.Errorf("getting charm hash: %w", err)
	}
```